private/bufpkg/bufconnect/interceptors.go

```diff
@@ -42,65 +43,136 @@ func NewSetCLIVersionInterceptor(version string) connect.UnaryInterceptorFunc {
 	return interceptor
 }
 
+// TokenFinder finds the token for NewAuthorizationInterceptorProvider
+type TokenFinder interface {
+	// RemoteToken returns the remote token from the remote address.
+	// setFromEnvVar is true if the returned token is from the tokenEnvKey environment variable.
+	RemoteToken(address string) (token string, setFromEnvVar bool)
+}
+
+type NetrcTokensProvider struct {
+	container         app.EnvContainer
+	getMachineForName func(app.EnvContainer, string) (netrc.Machine, error)
+}
+
+// NewNetrcTokensProvider returns a NetrcTokensProvider
+func NewNetrcTokensProvider(container app.EnvContainer, getMachineForName func(app.EnvContainer, string) (netrc.Machine, error)) *NetrcTokensProvider {
+	return &NetrcTokensProvider{container: container, getMachineForName: getMachineForName}
+}
+
+// RemoteToken finds the token from the .netrc file
+func (nt *NetrcTokensProvider) RemoteToken(address string) (string, bool) {
+	machine, err := nt.getMachineForName(nt.container, address)
+	if err != nil {
+		return "", false
+	}
+	if machine != nil {
+		return machine.Password(), false
+	}
+	return "", false
+}
+
+// TokenSetProvider is used to provide authentication token in NewAuthorizationInterceptorProvider
+type TokenSetProvider struct {
+	// true: the tokenSet is generated from environment variable tokenEnvKey
+	// false: otherwise
+	setBufTokenEnvVar bool
+	defaultToken      string
+	tokens            map[string]authKey
+}
+
+var _ TokenFinder = (*TokenSetProvider)(nil)
+
+// NewTokenSetProviderFromContainer creates a TokenSetProvider from the BUF_TOKEN environment variable
+func NewTokenSetProviderFromContainer(container app.EnvContainer) (*TokenSetProvider, error) {
+	bufToken := container.Env(tokenEnvKey)
+	tokenSet, err := NewTokenSetProviderFromString(bufToken)
+	if err != nil {
+		return nil, err
+	}
+	if bufToken != "" {
+		tokenSet.setBufTokenEnvVar = true
```

Comment:
If `bufToken == ""`, won't `RemoteToken` always return `""`? Whether or not a token came from an envVar feels like a property of a private constructor:



```go
func NewTokenProviderFromContainer(container app.EnvContainer) (TokenProvider, error) {
   return newTokenProviderFromString(container.Env(tokenEnvKey), true)
}

func NewTokenProviderFromString(token string) (TokenProvider, error) {
  return newTokenProviderFromString(token, false)
}

func newTokenProviderFromString(token string, isFromEnvVar bool) (TokenProvider, error) {
  ...
}
```